src/tiffcomposite_int.cpp

```diff
@@ -557,14 +541,18 @@ TiffComponent* TiffComponent::doAddChild(UniquePtr /*tiffComponent*/) {
 }  // TiffComponent::doAddChild
 
 TiffComponent* TiffDirectory::doAddChild(TiffComponent::UniquePtr tiffComponent) {
-  TiffComponent* tc = tiffComponent.release();
-  components_.push_back(tc);
+  TiffComponent* tc = tiffComponent.get();
+  components_.push_back(std::move(tiffComponent));
   return tc;
 }  // TiffDirectory::doAddChild
 
 TiffComponent* TiffSubIfd::doAddChild(TiffComponent::UniquePtr tiffComponent) {
-  auto d = dynamic_cast<TiffDirectory*>(tiffComponent.release());
-  ifds_.push_back(d);
+  auto d = dynamic_cast<TiffDirectory*>(tiffComponent.get());
+  if (!d) {
+    throw Error(ErrorCode::kerErrorMessage, "dynamic_cast to TiffDirectory failed");
+  }
+  tiffComponent.release();
+  ifds_.push_back(std::unique_ptr<TiffDirectory>(d));
```

Comment:
@neheb: do you know of a better way to do this? I need to dynamic_cast from `std::unique_ptr<TiffComponent>` to `std::unique_ptr<TiffDirectory>`.

Reply:
Nope.

dynamic casting is usually done with shared_ptr and std::dynamic_pointer_cast.

Effectively get() is being abused here to get C pointers.

Reply:
I just tested this.

replace push_back(std::unique_ptr()) with emplace_back

Reply:
I've replaced it with `emplace_back`.